source/sigv4.c

```diff
@@ -947,17 +1919,685 @@ static SigV4Status_t verifySigV4Parameters( const SigV4Parameters_t * pParams )
+                                                           size_t * pSignedHeadersLen )
+{
+    SigV4Status_t returnStatus = SigV4Success;
+    const char * pPath = NULL;
+    size_t pathLen = 0U;
+
+    if( returnStatus == SigV4Success )
+    {
+        pCanonicalContext->pBufCur = pCanonicalContext->pBufProcessing;
+        pCanonicalContext->bufRemaining = SIGV4_PROCESSING_BUFFER_LENGTH;
+
+        /* Write the HTTP Request Method to the canonical request. */
+        returnStatus = writeLineToCanonicalRequest( pParams->pHttpParameters->pHttpMethod,
+                                                    pParams->pHttpParameters->httpMethodLen,
+                                                    pCanonicalContext );
+    }
+
+    /* Set defaults for path and algorithm. */
+    if( ( pParams->pHttpParameters->pPath == NULL ) ||
+        ( pParams->pHttpParameters->pathLen == 0U ) )
+    {
+        /* If the absolute path is empty, use a forward slash (/). */
+        pPath = HTTP_EMPTY_PATH;
+        pathLen = HTTP_EMPTY_PATH_LEN;
+    }
+    else
+    {
+        pPath = pParams->pHttpParameters->pPath;
+        pathLen = pParams->pHttpParameters->pathLen;
+    }
+
+    if( returnStatus == SigV4Success )
+    {
+        /* Write the URI to the canonical request. */
+        if( pParams->pHttpParameters->flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG )
+        {
+            /* URI is already canonicalized, so just write it to the buffer as is. */
+            returnStatus = writeLineToCanonicalRequest( pPath,
+                                                        pathLen,
+                                                        pCanonicalContext );
+        }
+        else if( ( pParams->serviceLen == S3_SERVICE_NAME_LEN ) &&
+                 ( strncmp( pParams->pService, S3_SERVICE_NAME, S3_SERVICE_NAME_LEN ) == 0 ) )
+        {
+            /* S3 is the only service in which the URI must only be encoded once. */
+            returnStatus = generateCanonicalURI( pPath, pathLen, false, pCanonicalContext );
+        }
+        else
+        {
+            returnStatus = generateCanonicalURI( pPath, pathLen, true, pCanonicalContext );
+        }
+    }
+
+    if( returnStatus == SigV4Success )
+    {
+        /* Write the query to the canonical request. */
+        if( pParams->pHttpParameters->flags & SIGV4_HTTP_QUERY_IS_CANONICAL_FLAG )
+        {
+            /* HTTP query is already canonicalized, so just write it to the buffer as is. */
+            returnStatus = writeLineToCanonicalRequest( pParams->pHttpParameters->pQuery,
+                                                        pParams->pHttpParameters->queryLen,
+                                                        pCanonicalContext );
+        }
+        else
+        {
+            returnStatus = generateCanonicalQuery( pParams->pHttpParameters->pQuery, pParams->pHttpParameters->queryLen, pCanonicalContext );
+        }
```

Comment:
`pParams->pHttpParameters->pQuery` may be NULL also. That case needs to be handled.
As mentioned in the AWS docs:
[https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html]
Add the canonical query string, followed by a newline character. If the request does not include a query string, use an empty string (essentially, a blank line).

Reply:
Thanks for catching this :) I've removed the checks and assertions on `pQuery` being `NULL`